bundle/config/mutator/run_as.go

```diff
@@ -23,12 +27,120 @@ func (m *setRunAs) Name() string {
 	return "SetRunAs"
 }
 
+// Resources that satisfy one of the following conditions:
+//  1. Allow to set run_as for the resources to a different user from the current
+//     deployment user. For example, jobs.
+//  2. Does not make sense for these resources to run_as a different user. We do not
+//     have plans to add platform side support for `run_as` for these resources.
+//     For example, experiments or registered models.
+var allowListForRunAs = []string{"jobs", "models", "registered_models", "experiments"}
+
+// Resources that do not allow setting a run_as identity to a different user but
+// have plans to add platform side support for `run_as` for these resources at
+// some point in the future. For example, pipelines, model serving endpoints or lakeview dashboards.
+//
+// We expect the allow list and the deny list to form mutually exclusive and exhaustive
+// sets of all resource types that are supported by DABs.
+var denyListForRunAs = []string{"pipelines", "model_serving_endpoints"}
+
+type errorUnsupportedResourceTypeForRunAs struct {
+	resourceType  string
+	resourceValue dyn.Value
+	currentUser   string
+	runAsUser     string
+}
+
+// TODO(6 March 2024): This error message is big. We should split this once
+// diag.Diagnostics is ready.
+// TODO(6 March 2024): Link the docs page describing run_as semantics in the error below
+// once the page is ready.
+func (e errorUnsupportedResourceTypeForRunAs) Error() string {
+	return fmt.Sprintf("%s are not supported when the current deployment user is different from the bundle's run_as identity. Please deploy as the run_as identity. List of supported resources: [%s]. Location of the unsupported resource: %s. Current identity: %s. Run as identity: %s", e.resourceType, strings.Join(allowListForRunAs, ", "), e.resourceValue.Location(), e.currentUser, e.runAsUser)
+}
+
+func getRunAsIdentity(runAs dyn.Value) (string, error) {
+	// Get service principal name and user name from run_as section
+	runAsSp, err := dyn.Get(runAs, "service_principal_name")
+	if err != nil && !dyn.IsNoSuchKeyError(err) {
+		return "", err
+	}
+	runAsUser, err := dyn.Get(runAs, "user_name")
+	if err != nil && !dyn.IsNoSuchKeyError(err) {
+		return "", err
+	}
+
+	sp, spIsDefined := runAsSp.AsString()
+	user, userIsDefined := runAsUser.AsString()
+
+	switch {
+	case spIsDefined && userIsDefined:
+		return "", fmt.Errorf("run_as section must specify exactly one identity. A service_principal_name %q is specified at %s. A user_name %q is defined at %s", sp, runAsSp.Location(), user, runAsUser.Location())
+	case spIsDefined:
+		return sp, nil
+	case userIsDefined:
+		return user, nil
+	default:
+		return "", nil
```

Comment:
Isn't this error case as well, when both are unspecified?

Reply:
I think it's fine to not error in this case, especially since we are now using the typed configuration.

Reply:
Added error case for this!